flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/nn/QuadTree.scala

```diff
@@ -0,0 +1,301 @@
+      * (minDist adopted from "Nearest Neighbors Queries" by N. Roussopoulos et al.)
+      *
+      * @param obj
+      * @param radius
+      * @return
+      */
+    def isNear(obj: Vector, radius: Double): Boolean = {
+      if (minDist(obj) < radius) {
+        true
+      } else {
+        false
+      }
+    }
+
+    def minDist(obj: Vector): Double = {
+      var minDist = 0.0
+      for (i <- 0 to obj.size - 1) {
+        if (obj(i) < center(i) - width(i) / 2) {
+          minDist += math.pow(obj(i) - center(i) + width(i) / 2, 2)
+        } else if (obj(i) > center(i) + width(i) / 2) {
+          minDist += math.pow(obj(i) - center(i) - width(i) / 2, 2)
+        }
+      }
+      minDist
+    }
+
+    def whichChild(obj: Vector): Int = {
+
+      var count = 0
+      for (i <- 0 to obj.size - 1) {
+        if (obj(i) > center(i)) {
+          count += Math.pow(2, obj.size -1 - i).toInt
+        }
+      }
+      count
+    }
+
+    def makeChildren() {
+      val centerClone = center.copy
+      val cPart = partitionBox(centerClone, width)
+      val mappedWidth = 0.5*width.asBreeze
+      children = cPart.map(p => new Node(p, mappedWidth.fromBreeze, null))
+
+    }
+
+    /**
+     *  Recursive function that partitions a n-dim box by taking the (n-1) dimensional
+     * plane through the center of the box keeping the n-th coordinate fixed,
+     * then shifting it in the n-th direction up and down
+     * and recursively applying partitionBox to the two shifted (n-1) dimensional planes.
+     *
+     * @param center
+     * @param width
+     * @return
+     *
+     */
+    def partitionBox(center: Vector, width: Vector): Seq[Vector] = {
+
+      def partitionHelper(box: Seq[Vector], dim: Int): Seq[Vector] = {
+        if (dim >= width.size) {
+          box
+        } else {
+          val newBox = box.flatMap {
+            vector =>
+              val (up, down) = (vector.copy, vector)
+              up.update(dim, up(dim) - width(dim) / 4)
+              down.update(dim, down(dim) + width(dim) / 4)
+
+              Seq(up,down)
+          }
+          partitionHelper(newBox, dim + 1)
+        }
+      }
+      partitionHelper(Seq(center), 0)
+    }
+  }
+
+
+  val root = new Node( ((minVec.asBreeze + maxVec.asBreeze)*0.5).fromBreeze,
+    (maxVec.asBreeze - minVec.asBreeze).fromBreeze, null)
+
+    /**
+     *   simple printing of tree for testing/debugging
+     */
+  def printTree(){
+    printTreeRecur(root)
+  }
+
+  def printTreeRecur(n:Node){
+    if(n.children != null) {
+      for (c <- n.children){
+        printTreeRecur(c)
+      }
+    }else{
+      println("printing tree: n.objects " + n.objects)
+    }
+  }
+
+  /**
+   * Recursively adds an object to the tree
+   * @param obj
+   */
+  def insert(obj:Vector){
+    insertRecur(obj,root)
+  }
+
+  private def insertRecur(obj:Vector,n:Node) {
+    if(n.children==null) {
+      if(n.objects.length < maxPerBox )
+      {
+        n.objects += obj
+      }
+
+      else{
```

Comment:
`else` should start in the same line as the closing curly braces of the if branch.